src/common/snippets/include/snippets/pass_manager.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+#include "openvino/pass/manager.hpp"
+#include "openvino/pass/pass.hpp"
+#include "openvino/pass/validate.hpp"
+#include <typeinfo>
+
+
+namespace ov {
+namespace snippets {
+namespace pass {
+/**
+ * @brief Manager is like ov::pass::Manager, but allows to insert new passes at arbitrary places in the pipeline
+ * @ingroup snippets
+ */
+class Manager : public ov::pass::Manager {
+public:
+    ~Manager() override = default;
+    using PassBase = ov::pass::PassBase;
+    using Validate = ov::pass::Validate;
+    /**
+    * @brief PassPosition describes a particular position in a transformation pipeline,
+     * where a new transformation should be inserted.
+     * @param pass_name name of the anchor pass, the new pass will be inserted before/after it.
+     * Empty pass_name could mean either beginning or the end of the pipeline depending on the `after` flag.
+     * No default value.
+     * @param after `true` if the new pass should be inserted before the anchor pass, `false` otherwise (default).
+     * If `pass_name` is empty, `true` means the end, and `false` - the beginning of the pipeline.
+     * @param pass_instance the number of the pass with matching `pass_name` to be considered as the anchor pass.
+     * 0 (default) means the first pass with `pass_name` will be considered as the anchor pass.
+    * @ingroup snippets
+    */
+    class PassPosition {
+        const std::string m_pass_name;
+        const size_t m_pass_instance;
+        const bool m_after;
+
+    public:
+        using pass_list_type = std::vector<std::shared_ptr<ov::pass::PassBase>>;
+        explicit PassPosition(std::string pass_name, bool after = false, size_t pass_instance = 0)
+        : m_pass_name(std::move(pass_name)), m_pass_instance(pass_instance), m_after(after) {
+            OPENVINO_ASSERT((!m_pass_name.empty()) || (m_pass_instance == 0),
+                            "Non-zero pass_instance is not allowed if pass_name is not provided");
+        }
+
+        pass_list_type::const_iterator get_insert_position(const pass_list_type& pass_list) const;
+    };
+    using PositionedPass = std::pair<PassPosition, std::shared_ptr<PassBase>>;
```

Comment:
I'd suggest to avoid `pair` using. We make simple structure for that with representative names of fields instead of `first` and `second`. What do you think?

Also we would use this structure as parameter for `register_pass_instance()`

Reply:
Sure, good idea, thanks)

Reply:
> Also we would use this structure as parameter for `register_pass_instance()`
I remember now, why I did it this way 🙃 
1. It is logical to keep signatures of semantically equivalent methods similar. For example, signatures of `register_pass`, `register_pass_instance` and `insert_pass_instance` in our case.
2. `register_pass` signature is "fixed" by the idea that we want to specialize the original `ov::pass::Manager::register_pass`. Somehow, we need to understand that the provided list of arguments should be split in two parts: the first one specifies position, and the second one should be used to construct the pass itself. The easiest (and probably the only) way to do so, is to check whether the first argument can be used to create `PassPosition`, so the rest can be forwarded to the pass constructor. This determines the `register_pass` signature, so the rest of the methods are bound to take `PassPosition` as a separate first argument